### PR TITLE
Rewards improvements - yearly rate decrease and proper epoch rewards based on actual duration

### PR DESCRIPTION
**Pull request by igor-aptos**
### Description

- fix so that shorter epochs (due to governance proposals) give proportionally smaller rewards
- implemented yearly rate decrease, 1.5% per year, down to 3.5%

- Did a cleanup with it - to have yearly reward rate on chain, instead of per epoch - so that epoch duration change doesn't require changes to other configs.

### Test Plan
Current tests still passing. Will add few more tests.

### CODE REVIEW

**Review by gregnazario — COMMENTED** [review]
There's a lot of stuff here.  Have you tested it with the current mainnet / testnet genesis to ensure that nothing breaks?

**Review by wrwg — COMMENTED** [review]
Drive by comments. This is super critical code, so this likely requires:

- revisit specs and verification of staking @junkil-park 
- re-audit?
- AIP seems to be not needed? Design doc? (Not asking, just wondering)

**Review by davidiw — COMMENTED** [review]
could we maybe break this up into smaller commits / PRs... some of them can get in easily -- others likely require some auditing / deeper review.

**Comment by igor-aptos** [comment]
Let me split up the genesis cleanup from the actual code change. The rest I think is simpler to be a single PR, but let me know if you think otherwise after I resend it. 

This would require governance - to update the modules, and to set staking config values. Can that be done atomically in the single governance? @movekevin

**Comment by igor-aptos** [comment]
Split out genesis cleanup into #5780, and I stacked this PR on top of it.

Only meaningful code change here is in the move files - where we make the change on how much rewards to give. Rewards are controlled via new config, and uses new formula - to support two changes - reward based on actual epoch duration (i.e. in case of governance shortening the epoch), and yearly decreasing the rewards rate.

`rewards_apy_percentage: u64 = 7`
gets changed into 
```
  initial_yearly_rewards_rate: f64 = 0.07,
  min_yearly_rewards_rate: f64 = 0.035,
  yearly_rewards_rate_decrease: f64 = 0.985,
```
(in configuration, and in move denominators are added)

It would be good to get comments/review on those, while in parallel I add tests (both move tests, as well as end-to-end smoke test for the upgrade).

**Comment by movekevin** [comment]
> Let me split up the genesis cleanup from the actual code change. The rest I think is simpler to be a single PR, but let me know if you think otherwise after I resend it.
> 
> This would require governance - to update the modules, and to set staking config values. Can that be done atomically in the single governance? @movekevin

Yes it can be done in a single proposal with multi-step proposal support but I think we should still separate the feature flag enabling from the core code upgrade. This would help keep binary releases (which should mostly be no-op since all major changes should be gated) separate from enabling major behavior changes.

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move [hunk 1]
```diff
@@ -35,24 +46,36 @@ module aptos_framework::staking_config {
         allow_validator_set_change: bool,
         // The maximum rewards given out every epoch. This will be divided by the rewards rate denominator.
         // For example, 0.001% (0.00001) can be represented as 10 / 1000000.
-        rewards_rate: u64,
-        rewards_rate_denominator: u64,
+        _rewards_rate: u64,
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/staking_config.move:49]: You can't change the struct names unfortunately as that would fail the compatibility test (upgrade would fail). You can leave comments to document that these should be deprecated.

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move [hunk 2]
```diff
@@ -35,24 +46,36 @@ module aptos_framework::staking_config {
...
         // This is necessary to prevent a massive amount of new stake from joining that can potentially take down the
         // network if corresponding validators are not ready to participate in consensus in time.
         // This value is within (0, 50%), not inclusive.
         voting_power_increase_limit: u64,
     }
 
+    struct StakingRewardsConfig has copy, drop, key {
+        yearly_rewards_rate: u64,
+        min_yearly_rewards_rate: u64,
+
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/staking_config.move:61]: nit: Remove unnecessary empty lines.

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move [hunk 3]
```diff
@@ -35,24 +46,36 @@ module aptos_framework::staking_config {
...
+        _rewards_rate: u64,
+        _rewards_rate_denominator: u64,
         // Only this % of current total voting power is allowed to join the validator set in each epoch.
         // This is necessary to prevent a massive amount of new stake from joining that can potentially take down the
         // network if corresponding validators are not ready to participate in consensus in time.
         // This value is within (0, 50%), not inclusive.
         voting_power_increase_limit: u64,
     }
 
+    struct StakingRewardsConfig has copy, drop, key {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/staking_config.move:58]: Can you add some docs for this new struct and its fields?

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move [hunk 4]
```diff
@@ -35,24 +46,36 @@ module aptos_framework::staking_config {
...
+        yearly_rewards_rate: u64,
+        min_yearly_rewards_rate: u64,
+
+        rewards_rate_denominator: u64,
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/staking_config.move:62]: yearly_rewards_rate_denominator?

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move [hunk 5]
```diff
@@ -35,24 +46,36 @@ module aptos_framework::staking_config {
...
+
+        last_rate_decrease_time: u64,
+
+        yearly_rewards_rate_decrease_numerator: u64,
+        yearly_rewards_rate_decrease_denominator: u64,
+
+        year_in_micros: u64,
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/staking_config.move:69]: Don't we already have ONE_YEAR_IN_MICROS?

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move [hunk 6]
```diff
@@ -35,24 +46,36 @@ module aptos_framework::staking_config {
...
 
+    struct StakingRewardsConfig has copy, drop, key {
+        yearly_rewards_rate: u64,
+        min_yearly_rewards_rate: u64,
+
+        rewards_rate_denominator: u64,
+
+        last_rate_decrease_time: u64,
+
+        yearly_rewards_rate_decrease_numerator: u64,
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/staking_config.move:66]: Do we need this if the rewards rate decrease is likely 1.5% or along that line? We can always have the denominator be 10000 as it likely would not have so many trailing decimals like the current actual per-epoch or per-sec reward rate. With denominator = 10k, yearly decrease rate would become yearly_rewards_rate_bps (1.5% = 150 bps as 1 bps = 0.01%)

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move [hunk 7]
```diff
@@ -61,34 +84,58 @@ module aptos_framework::staking_config {
...
-        assert!(rewards_rate <= rewards_rate_denominator, error::invalid_argument(EINVALID_REWARDS_RATE));
-
         move_to(aptos_framework, StakingConfig {
             minimum_stake,
             maximum_stake,
             recurring_lockup_duration_secs,
             allow_validator_set_change,
-            rewards_rate,
-            rewards_rate_denominator,
+            _rewards_rate: 0,
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/staking_config.move:98]: Can you leave comments here explaining that this is set to 0 because it's been deprecated and replaced with the other struct?

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move [hunk 8]
```diff
@@ -16,9 +16,20 @@ module aptos_framework::staking_config {
...
+    /// Configured time period is not valid
+    const EINVALID_TIME_PERIOD: u64 = 7;
+    /// Time went backwards (between rate reconfig and current time)
+    const ETIME_WENT_BACKWARDS: u64 = 8;
 
     /// Limit the maximum value of `rewards_rate` in order to avoid any arithmetic overflow.
-    const MAX_REWARDS_RATE: u64 = 1000000;
+    const MAX_EPOCH_REWARDS_RATE: u64 = 100000;
+
+    const MAX_REWARDS_RATE: u64 = 100000000;
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/staking_config.move:29]: nit: Why move this here?

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move [hunk 9]
```diff
@@ -110,8 +157,39 @@ module aptos_framework::staking_config {
     }
 
     /// Return the reward rate.
-    public fun get_reward_rate(config: &StakingConfig): (u64, u64) {
-        (config.rewards_rate, config.rewards_rate_denominator)
+    public fun get_reward_rate(_config: &StakingConfig): (u64, u64) {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/staking_config.move:160]: This change should be gated behind a feature flag. And even if we have enabled per-second rewards rate, we can still actually return the per epoch rate here by doing the math based on per-sec rate. This would avoid breaking off chain systems that have not migrated to the new rewards rate getter function.

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move [hunk 10]
```diff
@@ -110,8 +157,39 @@ module aptos_framework::staking_config {
...
+        assert!(false, EDEPRECATED_FUNCTION);
+        (0, 1)
+    }
+
+    public fun get_epoch_reward_rate(epoch_duration: u64): (u64, u64) acquires StakingRewardsConfig {
+        let staking_rewards_config = borrow_global<StakingRewardsConfig>(@aptos_framework);
+        assert!(staking_rewards_config.year_in_micros > 0, EINVALID_TIME_PERIOD);
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/staking_config.move:167]: nit: Wrap EINVALID_TIME_PERIOD in error::invalid_argument

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move [hunk 11]
```diff
@@ -110,8 +157,39 @@ module aptos_framework::staking_config {
...
+        if (epoch_duration > 0) {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/staking_config.move:168]: I think we can skip this check since epoch duration cannot be 0.

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move [hunk 12]
```diff
@@ -110,8 +157,39 @@ module aptos_framework::staking_config {
...
+            (staking_rewards_config.yearly_rewards_rate * epoch_duration / staking_rewards_config.year_in_micros, staking_rewards_config.rewards_rate_denominator)
+        } else {
+            (0, 1)
+        }
+    }
+
+    public fun check_and_autodecrease_rewards_rate(current_time: u64) acquires StakingRewardsConfig {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/staking_config.move:175]: Two questions here:
1. Is this intentionally public so anyone can call this to update the rewards rate (no malicious behavior possible here since it's controlled based on last decrease time).
2. What's the benefit of saving the new rewards rate vs just computing it on the fly (initial_reward_time - (current_time - initial time) / year_in_secs * decrease_rate)?

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move [hunk 13]
```diff
@@ -147,29 +225,46 @@ module aptos_framework::staking_config {
         staking_config.recurring_lockup_duration_secs = new_recurring_lockup_duration_secs;
     }
 
+    public fun update_rewards_rate(
+        _aptos_framework: &signer,
+        _new_rewards_rate: u64,
+        _new_rewards_rate_denominator: u64,
+    ) {
+        assert!(false, EDEPRECATED_FUNCTION);
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/staking_config.move:233]: I think as a precaution, we might not want to deprecate this in case we need to change the old per-epoch rate for some reason. Wdyt?

File: aptos-move/framework/aptos-framework/sources/reconfiguration.move [hunk 14]
```diff
@@ -115,11 +120,17 @@ module aptos_framework::reconfiguration {
             return
         };
 
+        assert!(current_time > config_ref.last_reconfiguration_time, error::invalid_state(EINVALID_BLOCK_TIME));
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/reconfiguration.move:123]: Why do we need this check here? Isn't this already done in block::block_prologue?

File: aptos-move/framework/aptos-framework/sources/reconfiguration.move [hunk 15]
```diff
@@ -115,11 +120,17 @@ module aptos_framework::reconfiguration {
...
+        let epoch_duration = if (config_ref.last_reconfiguration_time == 0) {
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/reconfiguration.move:124]: Can you document that this is necessary or otherwise the validators would receive massive rewards in the network's first block? Are there other reasons I missed?

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move [hunk 16]
```diff
@@ -35,24 +46,36 @@ module aptos_framework::staking_config {
...
+        _rewards_rate_denominator: u64,
         // Only this % of current total voting power is allowed to join the validator set in each epoch.
         // This is necessary to prevent a massive amount of new stake from joining that can potentially take down the
         // network if corresponding validators are not ready to participate in consensus in time.
         // This value is within (0, 50%), not inclusive.
         voting_power_increase_limit: u64,
     }
 
+    struct StakingRewardsConfig has copy, drop, key {
+        yearly_rewards_rate: u64,
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/configs/staking_config.move:59]: Why not directly store per second reward rate?

File: aptos-move/framework/aptos-framework/sources/vesting.move [hunk 17]
```diff
@@ -1245,7 +1245,8 @@ module aptos_framework::vesting {
         commission = with_rewards(commission) + commission_on_staker_rewards;
         distribute(contract_address);
         // Rounding error leads to a dust amount of 1 transferred to the staker.
-        assert!(coin::balance<AptosCoin>(shareholder_address) == staker_rewards + 1, 0);
+        let shareholder_balance = coin::balance<AptosCoin>(shareholder_address);
```
Comment (movekevin) [inline aptos-move/framework/aptos-framework/sources/vesting.move:1248]: Unrelated refactoring?

**Review by movekevin — CHANGES_REQUESTED** [review]
Amazing! This looks good high level wise :)

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move (hunk 12 above)
Comment (igor-aptos) [reply aptos-move/framework/aptos-framework/sources/configs/staking_config.move:175]: 1. not intentionally public, what should I make it? friend?
2. three reasons:
  - I don't think we have "initial time". 
  - if at some point we want to change decrease rate, do we want it to retroactrively affect things (doubt it?) or only in the future? 
  - it is really an exponential formula ( * 0.985 every year), not a linear one, though we can make it be linear one instead

File: aptos-move/framework/aptos-framework/sources/configs/staking_config.move (hunk 16 above)
Comment (movekevin) [reply aptos-move/framework/aptos-framework/sources/configs/staking_config.move:59]: Yeah I was thinking this would be simpler since you would not need to divide by ONE_YEAR_IN_MICROS every time
